File: android/app/src/main/java/uz/ifcoder/operator/data/TokenStore.kt
package uz.ifcoder.operator.data

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/** Auth tokenni shifrlangan holda saqlaydi (SharedPreferences ustida EncryptedSharedPreferences). */
class TokenStore(context: Context) {

    private val prefs: SharedPreferences

    init {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        prefs = EncryptedSharedPreferences.create(
            context,
            "ifcoder_operator_secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    fun saveToken(token: String) {
        prefs.edit().putString(KEY_TOKEN, token).apply()
    }

    fun getToken(): String? = prefs.getString(KEY_TOKEN, null)

    fun isLoggedIn(): Boolean = !getToken().isNullOrBlank()

    fun clear() {
        prefs.edit().clear().apply()
    }

    /** FCM token onNewToken() chaqirilganda kelib, hali login qilinmagan bo'lsa vaqtincha saqlanadi. */
    fun savePendingFcmToken(token: String) {
        prefs.edit().putString(KEY_PENDING_FCM, token).apply()
    }

    fun takePendingFcmToken(): String? {
        val token = prefs.getString(KEY_PENDING_FCM, null)
        if (token != null) prefs.edit().remove(KEY_PENDING_FCM).apply()
        return token
    }

    companion object {
        private const val KEY_TOKEN = "auth_token"
        private const val KEY_PENDING_FCM = "pending_fcm_token"
    }
}

File: mobile/lib/data/token_store.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// Auth tokeni va login-oldi FCM tokenini saqlaydi. Native ilovadagi
/// TokenStore'ga mos — token doim shifrlangan (EncryptedSharedPreferences/Keychain)
/// holatda saqlanadi.
class TokenStore {
  TokenStore._();
  static final TokenStore instance = TokenStore._();

  final _storage = const FlutterSecureStorage(aOptions: AndroidOptions());

  static const _keyToken = 'auth_token';
  static const _keyPendingFcm = 'pending_fcm_token';

  Future<String?> getToken() => _storage.read(key: _keyToken);

  Future<void> saveToken(String token) => _storage.write(key: _keyToken, value: token);

  Future<bool> isLoggedIn() async => (await getToken()) != null;

  Future<void> clear() => _storage.delete(key: _keyToken);

  /// Login qilinmagan holda kelgan FCM tokenini vaqtincha saqlaydi —
  /// login muvaffaqiyatli bo'lgach serverga ro'yxatdan o'tkaziladi.
  Future<void> savePendingFcmToken(String token) => _storage.write(key: _keyPendingFcm, value: token);

  Future<String?> takePendingFcmToken() async {
    final value = await _storage.read(key: _keyPendingFcm);
    if (value != null) {
      await _storage.delete(key: _keyPendingFcm);
    }
    return value;
  }
}
